enhancements/rhcos/multi-stream.md

```diff
@@ -0,0 +1,467 @@
+valid values `rhcos-9`, `rhcos-10` for OCP and `scos-9` and `scos-10` for OKD.
+
+** Need to dig into OKD integration options **
+
+MachineConfigPool adds MachineOSStream as a field defaults to `rhcos-9` or `scos-9`
+to handle upgrades from releases without the field. Expectation is that
+installer always sets the value at install time going forward.
+
+data/data/rhcos.json is moved to data/data/coreos/rhcos-9.json
+data/data/fcos.json is moved to data/data/coreos/scos-9.json
+future rhcos-10.json and scos-10.json will be located in the same directory
+
+The release payload adds additional images for new streams and their
+extensions. Probably a good idea to tag osStream images uniformly and include
+stream name in the off chance that we ever grow to more than two streams.
+Whatever magic that plumbs in those image refs to MCPs needs to get updated.
+TODO: This is part that I (Scott) don't understand very well so need help.
+
+CoreOS stream files metadata section adds a `labels` array with at least
+`node.openshift.io/ID` and `node.openshift.io/version_ID` where the values
+match those fields in os-release metadata. Whenever nodes and machines are
+derived from these streams the manager for those resources should apply all
+labels from the stream metadata.
+
+** Note, it doesn't actually look like any of the current values in /etc/os-release
+are valid for to name specifically a major OS, do we need to add any? **
+
+```
+sh-5.1# cat /etc/os-release 
+NAME="Red Hat Enterprise Linux CoreOS"
+ID="rhcos"
+ID_LIKE="rhel fedora"
+VERSION="418.94.202409201939-0"
+VERSION_ID="4.18"
+VARIANT="CoreOS"
+VARIANT_ID=coreos
+PLATFORM_ID="platform:el9"
+PRETTY_NAME="Red Hat Enterprise Linux CoreOS 418.94.202409201939-0"
+ANSI_COLOR="0;31"
+CPE_NAME="cpe:/o:redhat:enterprise_linux:9::baseos::coreos"
+HOME_URL="https://www.redhat.com/"
+DOCUMENTATION_URL="https://docs.okd.io/latest/welcome/index.html"
+BUG_REPORT_URL="https://access.redhat.com/labs/rhir/"
+REDHAT_BUGZILLA_PRODUCT="OpenShift Container Platform"
+REDHAT_BUGZILLA_PRODUCT_VERSION="4.18"
+REDHAT_SUPPORT_PRODUCT="OpenShift Container Platform"
+REDHAT_SUPPORT_PRODUCT_VERSION="4.18"
+OPENSHIFT_VERSION="4.18"
+RHEL_VERSION=9.4
+OSTREE_VERSION="418.94.202409201939-0"
+```
+
+
+CoreOS stream files metadata section adds an 'eol' boolean field. Whenever the
+MCO encounters a MachineConfigPool which is configured with an osStream which
+is marked as `eol` the MCO should mark itself as Upgradeable=False with an
+appropriate message inhibiting upgrades to the next OCP minor version. In the
+unlikely event that plans change a subsequent z-stream update could be applied
+altering the `eol` value and MCO should then set Upgradeable=True whenever all
+currently in use streams are eol false.
+
+When the installer is built for OCP valid `osStream` must start with "rhcos"
+and match the name of a file in data/data/coreos/
+
+Somewhere in MCO's templates/ add streams/{rhcos-9,rhcos-10} anything outside
```

Comment:
What I suspect we may need to add to the MCO is a conditional like this for *user provided* machineconfig.

Reply:
If we hold that stream is configured at MCP level and that value is effectively immutable, wouldn't they just provide MachineConfig that matches the labels for each pool?

Need to vet the viability of stream being immutable with how we would potentially handle in-place major OS upgrades. My thinking for now is that entails moving a node from one pool to another rather than reconfiguring a pool for a different stream.